com.hazelcast.jet.JobStatusListenerTest.testLightListener_jobFails[onClient] [HZ-2195]

_master_ (commit fa2866756a08191e8fffc39f5755470236722652)

Failed on oracle-8 (little-endian): https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-nightly-little-endian/18/testReport/com.hazelcast.jet/JobStatusListenerTest/testLightListener_jobFails_onClient_/
<details><summary>Stacktrace:</summary>

```
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:911)
	at java.util.ArrayList$Itr.next(ArrayList.java:861)
	at com.hazelcast.test.HazelcastTestSupport.assertIterableEquals(HazelcastTestSupport.java:1001)
	at com.hazelcast.jet.JobStatusListenerTest.lambda$assertIterableEqualsEventually$16(JobStatusListenerTest.java:325)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1277)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1381)
	at com.hazelcast.jet.JobStatusListenerTest.assertIterableEqualsEventually(JobStatusListenerTest.java:325)
	at com.hazelcast.jet.JobStatusListenerTest.lambda$testLightListener_jobFails$10(JobStatusListenerTest.java:241)
	at com.hazelcast.jet.JobStatusListenerTest.testLightListener(JobStatusListenerTest.java:314)
	at com.hazelcast.jet.JobStatusListenerTest.testLightListener_jobFails(JobStatusListenerTest.java:232)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:750)

```
</details>


<details><summary>Standard output:</summary>

```
Finished Running Test: testListener_restartOnException[onClient] in 4.259 seconds.
Started Running Test: testLightListener_jobFails[onClient]
23:47:41,976 TRACE || - [ClientListenerService] hz.client_1918.eventRegistration- - hz.client_1918 [dev] [5.3.0-SNAPSHOT] Register attempt of ClientListenerRegistration{codec=com.hazelcast.jet.impl.ClientJobProxy$1@63ff6050, handler=com.hazelcast.jet.impl.ClientJobProxy$ClientJobStatusEventHandler@571affa4} to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2023-03-17 23:47:41.972, lastWriteTime=2023-03-17 23:47:41.976, closedTime=never}}
23:47:41,977 TRACE || - [ClientListenerService] hz.client_1918.eventRegistration- - hz.client_1918 [dev] [5.3.0-SNAPSHOT] Registered ClientListenerRegistration{codec=com.hazelcast.jet.impl.ClientJobProxy$1@63ff6050, handler=com.hazelcast.jet.impl.ClientJobProxy$ClientJobStatusEventHandler@571affa4} to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2023-03-17 23:47:41.977, lastWriteTime=2023-03-17 23:47:41.976, closedTime=never}}
23:47:41,977 TRACE || - [ClientListenerService] hz.client_1918.eventRegistration- - hz.client_1918 [dev] [5.3.0-SNAPSHOT] Register attempt of ClientListenerRegistration{codec=com.hazelcast.jet.impl.ClientJobProxy$1@63ff6050, handler=com.hazelcast.jet.impl.ClientJobProxy$ClientJobStatusEventHandler@571affa4} to MockedClientConnection{localAddress=[127.0.0.1]:40002, super=ClientConnection{alive=true, connectionId=2, channel=null, remoteAddress=[127.0.0.1]:5702, lastReadTime=2023-03-17 23:47:41.974, lastWriteTime=2023-03-17 23:47:41.972, closedTime=never}}
23:47:41,977 DEBUG || - [Planner] hz.JobStatusListenerTest_pensive_curie.client.thread-1 - Watermarks in the pipeline will be throttled to 1000
23:47:41,977 TRACE || - [ClientListenerService] hz.client_1918.eventRegistration- - hz.client_1918 [dev] [5.3.0-SNAPSHOT] Registered ClientListenerRegistration{codec=com.hazelcast.jet.impl.ClientJobProxy$1@63ff6050, handler=com.hazelcast.jet.impl.ClientJobProxy$ClientJobStatusEventHandler@571affa4} to MockedClientConnection{localAddress=[127.0.0.1]:40002, super=ClientConnection{alive=true, connectionId=2, channel=null, remoteAddress=[127.0.0.1]:5702, lastReadTime=2023-03-17 23:47:41.977, lastWriteTime=2023-03-17 23:47:41.977, closedTime=never}}
23:47:41,979 DEBUG || - [LightMasterContext] hz.JobStatusListenerTest_pensive_curie.client.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Start executing light job 0990-bd07-4108-0002, execution graph in DOT format:
digraph DAG {
	"itemStream" [localParallelism=1];
	"noop" [localParallelism=1];
	"itemStream" -> "noop" [queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
23:47:41,979 DEBUG || - [LightMasterContext] hz.JobStatusListenerTest_pensive_curie.client.thread-1 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Building execution plan for 0990-bd07-4108-0002
23:47:41,979 DEBUG || - [LightMasterContext] hz.JobStatusListenerTest_pensive_curie.cached.thread-7 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Built execution plans for 0990-bd07-4108-0002
23:47:41,979 DEBUG || - [InitExecutionOperation] hz.JobStatusListenerTest_pensive_curie.cached.thread-7 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Initializing execution plan for job 0990-bd07-4108-0002, execution 0990-bd07-4108-0002 from [127.0.0.1]:5701
23:47:41,980 DEBUG || - [JobExecutionService] hz.JobStatusListenerTest_pensive_curie.cached.thread-7 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Execution plan for light job ID=0990-bd07-4108-0002, jobName='0990-bd07-4108-0002', executionId=0990-bd07-4108-0002 initialized, will start the execution
23:47:41,981 DEBUG || - [InitExecutionOperation] hz.JobStatusListenerTest_eloquent_curie.generic-operation.thread-1 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Initializing execution plan for job 0990-bd07-4108-0002, execution 0990-bd07-4108-0002 from [127.0.0.1]:5701
23:47:41,982 DEBUG || - [JobExecutionService] hz.JobStatusListenerTest_eloquent_curie.cached.thread-12 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Execution plan for light job ID=0990-bd07-4108-0002, jobName='0990-bd07-4108-0002', executionId=0990-bd07-4108-0002 initialized, will start the execution
23:47:41,982 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-4 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Completed execution of job '0990-bd07-4108-0002', execution 0990-bd07-4108-0002
23:47:41,982 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-4 - [127.0.0.1]:5702 [dev] [5.3.0-SNAPSHOT] Execution of job '0990-bd07-4108-0002', execution 0990-bd07-4108-0002 completed
23:47:42,019 DEBUG || - [JobRepository] hz.JobStatusListenerTest_pensive_curie.cached.thread-7 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Job cleanup took 0ms
23:47:43,985  INFO || - [TaskletExecutionService] hz.JobStatusListenerTest_pensive_curie.jet.blocking.thread-0 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}
java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.JobStatusListenerTest.lambda$testLightListener_jobFails$648ca993$1(JobStatusListenerTest.java:232) ~[test-classes/:?]
	at com.hazelcast.jet.pipeline.test.TestSources$ItemStreamSource.fillBuffer(TestSources.java:236) ~[classes/:?]
	at com.hazelcast.function.BiConsumerEx.accept(BiConsumerEx.java:47) ~[classes/:?]
	at com.hazelcast.jet.impl.connector.ConvenientSourceP.complete(ConvenientSourceP.java:119) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:541) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:421) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:291) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:315) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) [?:1.8.0_351]
23:47:43,988 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Completed execution of job '0990-bd07-4108-0002', execution 0990-bd07-4108-0002
23:47:43,988 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Execution of job '0990-bd07-4108-0002', execution 0990-bd07-4108-0002 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:783) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_351]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_351]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:72) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:498) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.dismissTasklet(TaskletExecutionService.java:429) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:415) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:895) ~[?:1.8.0_351]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:369) ~[classes/:?]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_351]
Caused by: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.impl.execution.TaskletExecutionService.handleTaskletExecutionError(TaskletExecutionService.java:286) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService.access$600(TaskletExecutionService.java:80) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:325) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	... 1 more
Caused by: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.JobStatusListenerTest.lambda$testLightListener_jobFails$648ca993$1(JobStatusListenerTest.java:232) ~[test-classes/:?]
	at com.hazelcast.jet.pipeline.test.TestSources$ItemStreamSource.fillBuffer(TestSources.java:236) ~[classes/:?]
	at com.hazelcast.function.BiConsumerEx.accept(BiConsumerEx.java:47) ~[classes/:?]
	at com.hazelcast.jet.impl.connector.ConvenientSourceP.complete(ConvenientSourceP.java:119) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:541) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:421) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:291) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:315) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	... 1 more
23:47:43,993 ERROR || - [InitExecutionOperation] ForkJoinPool.commonPool-worker-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
com.hazelcast.jet.JetException: Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.impl.execution.TaskletExecutionService.handleTaskletExecutionError(TaskletExecutionService.java:286) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService.access$600(TaskletExecutionService.java:80) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:325) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_351]
Caused by: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.JobStatusListenerTest.lambda$testLightListener_jobFails$648ca993$1(JobStatusListenerTest.java:232) ~[test-classes/:?]
	at com.hazelcast.jet.pipeline.test.TestSources$ItemStreamSource.fillBuffer(TestSources.java:236) ~[classes/:?]
	at com.hazelcast.function.BiConsumerEx.accept(BiConsumerEx.java:47) ~[classes/:?]
	at com.hazelcast.jet.impl.connector.ConvenientSourceP.complete(ConvenientSourceP.java:119) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:541) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:421) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:291) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:315) ~[classes/:?]
	... 5 more
23:47:43,995 ERROR || - [SubmitJobOperation] hz.JobStatusListenerTest_pensive_curie.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.3.0-SNAPSHOT] Execution on a member failed: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
com.hazelcast.jet.JetException: Execution on a member failed: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.impl.LightMasterContext.findError(LightMasterContext.java:326) ~[classes/:?]
	at com.hazelcast.jet.impl.LightMasterContext.lambda$null$2(LightMasterContext.java:171) ~[classes/:?]
	at com.hazelcast.jet.impl.LightMasterContext.lambda$invokeOnParticipant$8(LightMasterContext.java:297) ~[classes/:?]
	at com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode.lambda$execute$0(AbstractInvocationFuture.java:1570) ~[classes/:?]
	at com.hazelcast.test.TestLoggingUtils$CustomTestNameAwareForkJoinPool$TestNameAwareRunnable.run(TestLoggingUtils.java:104) ~[test-classes/:?]
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1067) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1703) ~[?:1.8.0_351]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:172) ~[?:1.8.0_351]
Caused by: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{0990-bd07-4108-0002/itemStream#0}: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.impl.execution.TaskletExecutionService.handleTaskletExecutionError(TaskletExecutionService.java:286) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService.access$600(TaskletExecutionService.java:80) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:325) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_351]
Caused by: java.lang.ArithmeticException: / by zero
	at com.hazelcast.jet.JobStatusListenerTest.lambda$testLightListener_jobFails$648ca993$1(JobStatusListenerTest.java:232) ~[test-classes/:?]
	at com.hazelcast.jet.pipeline.test.TestSources$ItemStreamSource.fillBuffer(TestSources.java:236) ~[classes/:?]
	at com.hazelcast.function.BiConsumerEx.accept(BiConsumerEx.java:47) ~[classes/:?]
	at com.hazelcast.jet.impl.connector.ConvenientSourceP.complete(ConvenientSourceP.java:119) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:541) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:421) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:291) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:315) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_351]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_351]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_351]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_351]
23:47:45,320  INFO |testLightListener_jobFails[onClient]| - [SimpleTestInClusterSupport] Time-limited test - Ditching 17 jobs in SimpleTestInClusterSupport.@After: [0990-bcfb-3bc9-0002, 0990-bcfb-3bc1-0002, 0990-bcfb-3bc0-0002, 0990-bcfb-3bc3-0002, 0990-bcfb-3bc4-0002, 0990-bcfb-3bc6-0002, 0990-bd07-4104-0002, 0990-bd07-4107-0002, 0990-bd07-4106-0002, 0990-bd07-4101-0002, 0990-bd07-4100-0002, 0990-bd07-4103-0002, 0990-bcef-10c9-0001, 0990-bcef-10c7-0001, 0990-bcef-10c4-0001, 0990-bcef-10c5-0001, 0990-bcef-10c2-0001]
23:47:45,340  INFO |testLightListener_jobFails[onClient]| - [SimpleTestInClusterSupport] Time-limited test - Destroying 0 distributed objects in SimpleTestInClusterSupport.@After: []
BuildInfo right after testLightListener_jobFails[onClient](com.hazelcast.jet.JobStatusListenerTest): BuildInfo{version='5.3.0-SNAPSHOT', build='20230317', buildNumber=20230317, revision=fa28667, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testLightListener_jobFails[onClient](com.hazelcast.jet.JobStatusListenerTest):'
23:47:40, accumulated pauses: 1330 ms, max pause: 1287 ms, pauses over 1000 ms: 1
23:47:45, accumulated pauses: 0 ms, max pause: 0 ms, pauses over 1000 ms: 0


No metrics recorded during the test

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-nightly-little-endian/18/

--- a/hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java
@@ -531,7 +531,9 @@ public class MasterJobContext {
         if (jobStatus != NOT_RUNNING && jobStatus != STARTING && jobStatus != RUNNING) {
             throw new IllegalStateException("Restart scheduled in an unexpected state: " + jobStatus);
         }
-        mc.setJobStatus(NOT_RUNNING, description, false);
+        if (jobStatus != NOT_RUNNING) {
+            mc.setJobStatus(NOT_RUNNING, description, false);
+        }
         mc.coordinationService().scheduleRestart(mc.jobId());
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
@@ -51,6 +51,7 @@ import java.util.UUID;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.ExecutionException;
 import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Supplier;
@@ -266,17 +267,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     }
 
     protected void testListener(JobConfig config, Object source, BiConsumer<Job, JobStatusLogger> test) {
-        Pipeline p = Pipeline.create();
-        (source instanceof BatchSource
-                    ? p.readFrom((BatchSource<?>) source)
-                    : p.readFrom((StreamSource<?>) source).withoutTimestamps())
-                .writeTo(Sinks.noop());
-
-        Job job = instance.get().getJet().newJob(p, config);
-        JobStatusLogger listener = new JobStatusLogger(job);
-        jobIdString = job.getIdString();
-        registrationId = listener.registrationId;
-        test.accept(job, listener);
+        testListener(source, (jet, p) -> jet.newJob(p, config), test);
     }
 
     protected void testListener(Object source, BiConsumer<Job, JobStatusLogger> test) {
@@ -291,17 +282,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     }
 
     protected void testLightListener(Object source, BiConsumer<Job, JobStatusLogger> test) {
-        Pipeline p = Pipeline.create();
-        (source instanceof BatchSource
-                    ? p.readFrom((BatchSource<?>) source)
-                    : p.readFrom((StreamSource<?>) source).withoutTimestamps())
-                .writeTo(Sinks.noop());
-
-        Job job = instance.get().getJet().newLightJob(p);
-        JobStatusLogger listener = new JobStatusLogger(job);
-        jobIdString = job.getIdString();
-        registrationId = listener.registrationId;
-        test.accept(job, listener);
+        testListener(source, JetService::newLightJob, test);
     }
 
     protected void testLightListener(Object source, Consumer<Job> test, String log) {
@@ -309,6 +290,21 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
             test.accept(job);
             assertEqualsEventually(listener.log, log);
         });
+    }
+
+    private void testListener(Object source, BiFunction<JetService, Pipeline, Job> submit,
+                              BiConsumer<Job, JobStatusLogger> test) {
+        Pipeline p = Pipeline.create();
+        (source instanceof BatchSource
+                    ? p.readFrom((BatchSource<?>) source)
+                    : p.readFrom((StreamSource<?>) source).withoutTimestamps())
+                .writeTo(Sinks.noop());
+
+        Job job = submit.apply(instance.get().getJet(), p);
+        JobStatusLogger listener = new JobStatusLogger(job);
+        jobIdString = job.getIdString();
+        registrationId = listener.registrationId;
+        test.accept(job, listener);
     }
 
     @SafeVarargs

--- a/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
@@ -397,7 +397,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     static class FiniteStreamSource {
         final long periodNanos;
         final long maxSequence;
-        long emitSchedule = Long.MIN_VALUE;
+        long emitSchedule;
         long sequence;
 
         FiniteStreamSource(long period, TimeUnit unit, long maxSequence) {
@@ -407,6 +407,9 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
         void fillBuffer(SourceBuffer<Long> buf) {
             long nowNs = System.nanoTime();
+            if (emitSchedule == 0) {
+                emitSchedule = nowNs;
+            }
             if (nowNs >= emitSchedule) {
                 buf.add(sequence++);
                 emitSchedule += periodNanos;

--- a/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
@@ -238,7 +238,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
                         failure = e.getCause();
                     }
                     assertNotNull(failure);
-                    assertIterableEqualsEventually(listener.log,
+                    assertEqualsEventually(listener.log,
                             "Jet: RUNNING -> FAILED (" + failure + ")");
                 });
     }
@@ -317,12 +317,17 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     protected void testLightListener(Object source, Consumer<Job> test, String log) {
         testLightListener(source, (job, listener) -> {
             test.accept(job);
-            assertIterableEqualsEventually(listener.log, log);
+            assertEqualsEventually(listener.log, log);
         });
     }
 
-    protected static void assertIterableEqualsEventually(Iterable<?> actual, Object... expected) {
-        assertTrueEventually(() -> assertIterableEquals(actual, expected));
+    @SafeVarargs
+    protected static <T> void assertEqualsEventually(List<T> actual, T... expected) {
+        assertTrueEventually(() -> {
+            List<T> actualCopy = new ArrayList<>(actual);
+            assertEquals("length", actualCopy.size(), expected.length);
+            assertEquals(actualCopy, asList(expected));
+        });
     }
 
     @SafeVarargs

--- a/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/JobStatusListenerTest.java
@@ -18,8 +18,10 @@ package com.hazelcast.jet;
 
 import com.hazelcast.client.impl.spi.impl.listener.ClientListenerServiceImpl;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.function.ConsumerEx;
 import com.hazelcast.jet.config.JobConfig;
 import com.hazelcast.jet.core.JetTestSupport;
+import com.hazelcast.jet.core.Processor.Context;
 import com.hazelcast.jet.impl.JobEventService;
 import com.hazelcast.jet.pipeline.BatchSource;
 import com.hazelcast.jet.pipeline.Pipeline;
@@ -27,7 +29,7 @@ import com.hazelcast.jet.pipeline.Sinks;
 import com.hazelcast.jet.pipeline.SourceBuilder;
 import com.hazelcast.jet.pipeline.SourceBuilder.SourceBuffer;
 import com.hazelcast.jet.pipeline.StreamSource;
-import com.hazelcast.jet.pipeline.test.TestSources;
+import com.hazelcast.map.IMap;
 import com.hazelcast.test.HazelcastParametrizedRunner;
 import com.hazelcast.test.HazelcastSerialParametersRunnerFactory;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -48,7 +50,6 @@ import java.util.List;
 import java.util.UUID;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -59,7 +60,6 @@ import static com.hazelcast.jet.core.JobStatus.RUNNING;
 import static com.hazelcast.jet.core.JobStatus.SUSPENDED;
 import static com.hazelcast.spi.impl.eventservice.impl.EventServiceTest.getEventService;
 import static java.util.Arrays.asList;
-import static java.util.concurrent.TimeUnit.SECONDS;
 import static java.util.stream.Collectors.toList;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
@@ -71,7 +71,8 @@ import static org.junit.Assert.assertTrue;
 @Category(SlowTest.class)
 public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     private static final Function<String, String> SIMPLIFY = log -> log.replaceAll("(?<=\\().*: ", "");
-    private static final String MAP_NAME = "runCount";
+    private static final String ADVANCE_MAP_NAME = "advanceCount";
+    private static final String RUN_MAP_NAME = "runCount";
 
     @Parameters(name = "{0}")
     public static Iterable<Object[]> parameters() {
@@ -85,20 +86,14 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
         return new Object[] {name, supplier};
     }
 
-    /**
-     * Used to generate unique job ids to be used inside jobs, e.g. sinks,
-     * since it is not easy to access the context.
-     */
-    private static int nextJobId;
-
     @Parameter(0)
     public String mode;
 
     @Parameter(1)
     public Supplier<HazelcastInstance> instance;
 
-    private String jobIdString;
-    private UUID registrationId;
+    protected String jobIdString;
+    protected UUID registrationId;
 
     @BeforeClass
     public static void setup() {
@@ -107,7 +102,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
     @Test
     public void testListener_waitForCompletion() {
-        testListener(finiteStream(1, SECONDS, 2),
+        testListener(batchSource(),
                 Job::join,
                 "Jet: NOT_RUNNING -> STARTING",
                 "Jet: STARTING -> RUNNING",
@@ -116,7 +111,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
     @Test
     public void testListener_suspend_resume_restart_cancelJob() {
-        testListener(TestSources.itemStream(1, (t, s) -> s),
+        testListener(streamSource(),
                 (job, listener) -> {
                     assertJobStatusEventually(job, RUNNING);
                     job.suspend();
@@ -142,7 +137,9 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
     @Test
     public void testListener_jobFails() {
-        testListener(TestSources.itemStream(1, (t, s) -> 1 / (2 - s)),
+        testListener(batchSource(r -> {
+                    throw new JetException("mock error");
+                }),
                 (job, listener) -> {
                     Throwable failure = null;
                     try {
@@ -161,13 +158,9 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     @Test
     public void testListener_restartOnException() {
         testListener(new JobConfig().setAutoScaling(true),
-                TestSources.itemStream(1,
-                        (t, s) -> {
-                            if (s == 2) {
-                                throw new RestartableException();
-                            }
-                            return s;
-                        }),
+                streamSource(r -> {
+                    throw new RestartableException();
+                }),
                 (job, listener) -> {
                     assertEqualsEventually(listener::runCount, 2);
                     cancelAndJoin(job);
@@ -184,7 +177,9 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
     @Test
     public void testListener_suspendOnFailure() {
         testListener(new JobConfig().setSuspendOnFailure(true),
-                TestSources.itemStream(1, (t, s) -> 1 / (2 - s)),
+                batchSource(r -> {
+                    throw new JetException("mock error");
+                }),
                 (job, listener) -> {
                     assertJobStatusEventually(job, SUSPENDED);
                     String[] failure = new String[1];
@@ -201,7 +196,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
     @Test
     public void testListenerDeregistration() {
-        testListener(TestSources.itemStream(1, (t, s) -> s),
+        testListener(streamSource(),
                 (job, listener) -> {
                     assertJobStatusEventually(job, RUNNING);
                     listener.deregister();
@@ -215,21 +210,23 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
     @Test
     public void testLightListener_waitForCompletion() {
-        testLightListener(finiteStream(1, SECONDS, 2),
+        testLightListener(batchSource(),
                 Job::join,
                 "Jet: RUNNING -> COMPLETED");
     }
 
     @Test
     public void testLightListener_cancelJob() {
-        testLightListener(TestSources.itemStream(1, (t, s) -> s),
+        testLightListener(streamSource(),
                 Job::cancel,
                 "User: RUNNING -> FAILED (Cancel)");
     }
 
     @Test
     public void testLightListener_jobFails() {
-        testLightListener(TestSources.itemStream(1, (t, s) -> 1 / (2 - s)),
+        testLightListener(batchSource(r -> {
+                    throw new JetException("mock error");
+                }),
                 (job, listener) -> {
                     Throwable failure = null;
                     try {
@@ -245,7 +242,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
 
     @Test
     public void testLightListenerDeregistration() {
-        testLightListener(TestSources.itemStream(1, (t, s) -> s),
+        testLightListener(streamSource(),
                 (job, listener) -> {
                     listener.deregister();
                     job.cancel();
@@ -268,22 +265,15 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
         });
     }
 
-    /**
-     * @param source A {@link BatchSource BatchSource&lt;Long&gt;} or {@link StreamSource
-     *        StreamSource&lt;Long&gt;} where the first element is {@code 0L}, which is used
-     *        to increment the {@linkplain JobStatusLogger#runCount run count} of the job.
-     */
     protected void testListener(JobConfig config, Object source, BiConsumer<Job, JobStatusLogger> test) {
         Pipeline p = Pipeline.create();
-        int jobId = nextJobId++;
         (source instanceof BatchSource
-                    ? p.readFrom((BatchSource<Long>) source)
-                    : p.readFrom((StreamSource<Long>) source).withoutTimestamps())
-                .writeTo(Sinks.<Long, Integer, Integer>mapWithUpdating(
-                        MAP_NAME, s -> jobId, (i, s) -> s == 0 ? (i == null ? 0 : i) + 1 : i));
+                    ? p.readFrom((BatchSource<?>) source)
+                    : p.readFrom((StreamSource<?>) source).withoutTimestamps())
+                .writeTo(Sinks.noop());
 
         Job job = instance.get().getJet().newJob(p, config);
-        JobStatusLogger listener = new JobStatusLogger(job, jobId);
+        JobStatusLogger listener = new JobStatusLogger(job);
         jobIdString = job.getIdString();
         registrationId = listener.registrationId;
         test.accept(job, listener);
@@ -308,7 +298,7 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
                 .writeTo(Sinks.noop());
 
         Job job = instance.get().getJet().newLightJob(p);
-        JobStatusLogger listener = new JobStatusLogger(job, -1);
+        JobStatusLogger listener = new JobStatusLogger(job);
         jobIdString = job.getIdString();
         registrationId = listener.registrationId;
         test.accept(job, listener);
@@ -344,13 +334,12 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
         public final List<String> log = Collections.synchronizedList(new ArrayList<>());
         public final UUID registrationId;
         final Job job;
-        final int jobId;
         Thread originalThread;
 
-        JobStatusLogger(Job job, int jobId) {
+        public JobStatusLogger(Job job) {
             this.job = job;
-            this.jobId = jobId;
             registrationId = job.addStatusListener(this);
+            advance(1);
         }
 
         @Override
@@ -377,43 +366,73 @@ public class JobStatusListenerTest extends SimpleTestInClusterSupport {
          * {@link HazelcastTestSupport#assertEqualsEventually assertEqualsEventually(listener::runCount,
          * lastRunCount + 1)} can be used.
          */
-        int runCount() {
-            Integer count = (Integer) instance.get().getMap(MAP_NAME).get(jobId);
-            return count == null ? 0 : count;
+        public int runCount() {
+            return (int) instance.get().getMap(RUN_MAP_NAME).getOrDefault(job.getId(), 0);
         }
 
-        void deregister() {
+        public void advance(int count) {
+            JobStatusListenerTest.this.advance(job.getId(), count);
+        }
+
+        public void deregister() {
             job.removeStatusListener(registrationId);
         }
     }
 
-    protected static BatchSource<Long> finiteStream(long period, TimeUnit unit, long maxSequence) {
-        return SourceBuilder.batch("finiteStream",
-                        ctx -> new FiniteStreamSource(period, unit, maxSequence))
-                .fillBufferFn(FiniteStreamSource::fillBuffer)
+    protected void advance(long jobId, int count) {
+        instance.get().getMap(ADVANCE_MAP_NAME).put(jobId, count);
+    }
+
+    protected static BatchSource<Integer> batchSource() {
+        return batchSource(r -> { });
+    }
+
+    protected static BatchSource<Integer> batchSource(ConsumerEx<Integer> action) {
+        return SourceBuilder.batch("batchSource",
+                        ctx -> new TestSource(ctx, action, false))
+                .fillBufferFn(TestSource::fillBuffer)
                 .build();
     }
 
-    static class FiniteStreamSource {
-        final long periodNanos;
-        final long maxSequence;
-        long emitSchedule;
-        long sequence;
+    protected static BatchSource<Integer> streamSource() {
+        return streamSource(r -> { });
+    }
 
-        FiniteStreamSource(long period, TimeUnit unit, long maxSequence) {
-            periodNanos = unit.toNanos(period);
-            this.maxSequence = maxSequence;
+    protected static BatchSource<Integer> streamSource(ConsumerEx<Integer> action) {
+        return SourceBuilder.batch("streamSource",
+                        ctx -> new TestSource(ctx, action, true))
+                .fillBufferFn(TestSource::fillBuffer)
+                .build();
+    }
+
+    static class TestSource {
+        final Runnable incrementRunCount;
+        final Supplier<Boolean> advance;
+        final ConsumerEx<Integer> action;
+        final boolean streaming;
+        final int currentRun;
+        boolean firstRun = true;
+
+        TestSource(Context ctx, ConsumerEx<Integer> action, boolean streaming) {
+            IMap<Long, Integer> runCount = ctx.hazelcastInstance().getMap(RUN_MAP_NAME);
+            IMap<Long, Integer> allowedRunCount = ctx.hazelcastInstance().getMap(ADVANCE_MAP_NAME);
+            long jobId = ctx.jobId();
+            currentRun = runCount.getOrDefault(jobId, 0) + 1;
+            incrementRunCount = () -> runCount.compute(jobId, (id, count) -> count == null ? 1 : count + 1);
+            advance = () -> allowedRunCount.getOrDefault(jobId, 0) >= currentRun;
+            this.action = action;
+            this.streaming = streaming;
         }
 
-        void fillBuffer(SourceBuffer<Long> buf) {
-            long nowNs = System.nanoTime();
-            if (emitSchedule == 0) {
-                emitSchedule = nowNs;
+        void fillBuffer(SourceBuffer<Integer> buf) {
+            if (firstRun) {
+                incrementRunCount.run();
+                firstRun = false;
             }
-            if (nowNs >= emitSchedule) {
-                buf.add(sequence++);
-                emitSchedule += periodNanos;
-                if (sequence > maxSequence) {
+            if (advance.get()) {
+                action.accept(currentRun);
+                buf.add(currentRun);
+                if (!streaming) {
                     buf.close();
                 }
             }

